engine/runtime/src/main/scala/org/enso/interpreter/instrument/command/AttachVisualisationCmd.scala

```diff
@@ -22,11 +22,15 @@ class AttachVisualisationCmd(
     ctx: RuntimeContext,
     ec: ExecutionContext
   ): Future[Unit] = {
-    if (doesContextExist) {
+    val contextId = request.visualisationConfig.executionContextId
+    ctx.locking.acquireContextLock(contextId)
+    val result = if (doesContextExist) {
       attachVisualisation()
     } else {
       replyWithContextNotExistError()
     }
+    ctx.locking.releaseContextLock(contextId)
+    result
```

Comment:
This is unsafe and can deadlock. You always need to release a lock in a `finally` block to ensure that it's exception safe.